From 20c7eb8ccbb82b43918a0eb4fc1ea062182dd2ec Mon Sep 17 00:00:00 2001
From: zrj <rimvydas.jasinskas@gmail.com>
Date: Sun, 2 Feb 2020 12:07:45 +0200
Subject: [PATCH 06/22] stat(1): Disable -H support in btools.

 While there, limit features in btools too (used only in sys/boot/).
---
 usr.bin/stat/Makefile |  6 ++++--
 usr.bin/stat/stat.c   | 42 ++++++++++++++++++++++++++++--------------
 2 files changed, 32 insertions(+), 16 deletions(-)

diff --git a/usr.bin/stat/Makefile b/usr.bin/stat/Makefile
index 5820dae7c5..c9f321728d 100644
--- a/usr.bin/stat/Makefile
+++ b/usr.bin/stat/Makefile
@@ -1,9 +1,11 @@
 # $FreeBSD: head/usr.bin/stat/Makefile 216206 2010-12-05 21:33:05Z dougb $
 
 PROG=	stat
-CFLAGS+=-DHAVE_CONFIG_H=0
-
 LINKS=	${BINDIR}/stat ${BINDIR}/readlink
 MLINKS=	stat.1 readlink.1
 
+.if defined(BOOTSTRAPPING)
+CFLAGS+= -DBOOTSTRAPPING
+.endif
+
 .include <bsd.prog.mk>
diff --git a/usr.bin/stat/stat.c b/usr.bin/stat/stat.c
index 33870ea711..2318efd3d7 100644
--- a/usr.bin/stat/stat.c
+++ b/usr.bin/stat/stat.c
@@ -31,16 +31,6 @@
  * $FreeBSD: head/usr.bin/stat/stat.c 265893 2014-05-11 18:49:18Z thomas $
  */
 
-#if HAVE_CONFIG_H
-#include "config.h" 
-#else  /* HAVE_CONFIG_H */
-#define HAVE_STRUCT_STAT_ST_FLAGS 1
-#define HAVE_STRUCT_STAT_ST_GEN 1
-#define HAVE_STRUCT_STAT_ST_BIRTHTIME 0
-#define HAVE_STRUCT_STAT_ST_ATIM 1
-#define HAVE_DEVNAME 1
-#endif /* HAVE_CONFIG_H */
-
 #include <sys/param.h>
 #include <sys/types.h>
 #include <sys/stat.h>
@@ -59,6 +49,20 @@
 #include <time.h>
 #include <unistd.h>
 
+#ifdef BOOTSTRAPPING
+#define HAVE_DEVNAME 0
+#define HAVE_STRUCT_STAT_ST_FLAGS 0
+#define HAVE_STRUCT_STAT_ST_GEN 0
+#define HAVE_STRUCT_STAT_ST_BIRTHTIME 0
+#define HAVE_STRUCT_STAT_ST_ATIM 1
+#else
+#define HAVE_DEVNAME 1
+#define HAVE_STRUCT_STAT_ST_FLAGS 1
+#define HAVE_STRUCT_STAT_ST_GEN 1
+#define HAVE_STRUCT_STAT_ST_BIRTHTIME 0
+#define HAVE_STRUCT_STAT_ST_ATIM 1
+#endif
+
 #if HAVE_STRUCT_STAT_ST_FLAGS
 #define DEF_F "%#Xf "
 #define RAW_F "%f "
@@ -180,7 +184,9 @@ static int	format1(const struct stat *,	/* stat info */
 		    char *, size_t,		/* a place to put the output */
 		    int, int, int, int,		/* the parsed format */
 		    int, int);
+#if HAVE_DEVNAME
 static int	hex2byte(const char [2]);
+#endif
 #if HAVE_STRUCT_STAT_ST_FLAGS
 static char	*xfflagstostr(unsigned long);
 #endif
@@ -201,8 +207,10 @@ main(int argc, char *argv[])
 	int ch, rc, errs, am_readlink;
 	int lsF, fmtchar, usestat, nfs_handle, fn, nonl, quiet;
 	const char *statfmt, *options, *synopsis;
+#if HAVE_DEVNAME
 	char dname[sizeof _PATH_DEV + MNAMELEN] = _PATH_DEV;
 	fhandle_t fhnd;
+#endif
 	const char *file;
 
 	am_readlink = 0;
@@ -314,17 +322,20 @@ main(int argc, char *argv[])
 	errs = 0;
 	do {
 		if (argc == 0) {
+#if HAVE_DEVNAME
 			if (fdevname_r(STDIN_FILENO, dname +
 			    sizeof _PATH_DEV - 1, MNAMELEN) == 0)
 				file = dname;
 			else
+#endif
 				file = "(stdin)";
 			rc = fstat(STDIN_FILENO, &st);
 		} else {
-			int j;
-
 			file = argv[0];
 			if (nfs_handle) {
+#if HAVE_DEVNAME
+				int j;
+
 				rc = 0;
 				bzero(&fhnd, sizeof(fhnd));
 				j = MIN(2 * sizeof(fhnd), strlen(file));
@@ -343,7 +354,9 @@ main(int argc, char *argv[])
 					errno = EINVAL;
 				else
 					rc = fhstat(&fhnd, &st);
-
+#else
+				err(1, "-H is disabled in btools");
+#endif
 			} else if (usestat) {
 				/*
 				 * Try stat() and if it fails, fall back to
@@ -1082,7 +1095,7 @@ format1(const struct stat *st,
 	return (snprintf(buf, blen, lfmt, data));
 }
 
-
+#if HAVE_DEVNAME
 #define hex2nibble(c) (c <= '9' ? c - '0' : toupper(c) - 'A' + 10)
 static int
 hex2byte(const char c[2]) {
@@ -1090,3 +1103,4 @@ hex2byte(const char c[2]) {
 		return -1;
 	return (hex2nibble(c[0]) << 4) + hex2nibble(c[1]);
 }
+#endif
-- 
2.25.0

